File: src/presentation/portraits.ts
import type { VisualReference } from '../core/index.js';

function metadata(reference: VisualReference): string {
  let text = `${reference.caption} ${reference.id} ${reference.url}`;
  try {
    text = decodeURIComponent(text);
  } catch {
    /* Keep readable source metadata. */
  }
  return text.replace(/[_-]/g, ' ');
}

/** Source labels identify coverage, not pixel analysis. */
export function isFullBodyReference(reference: VisualReference): boolean {
  return /\b(full\s*body|whole\s*body|entire\s*body|turnaround|character sheet)\b/i.test(
    metadata(reference),
  );
}

export function safeReference(reference: VisualReference): boolean {
  return [reference.url, reference.sourceUrl].every((value) => {
    try {
      const url = new URL(value);
      return ['http:', 'https:'].includes(url.protocol) && !url.username && !url.password;
    } catch {
      return false;
    }
  });
}

/** Prefer portrait-labelled art over action and full-body references, without character tables. */
export function rankedPortraits(references: VisualReference[]): VisualReference[] {
  const score = (reference: VisualReference) => {
    const text = metadata(reference);
    return (
      (reference.kind === 'appearance' ? 15 : 0) +
      (/\b(portrait|headshot|close\s*up|bust)\b/i.test(text) ? 80 : 0) +
      (/\b(infobox|profile)\b/i.test(text) ? 35 : 0) +
      (/\banime\b/i.test(text) ? 10 : 0) -
      (reference.kind === 'pose' ? 25 : 0) -
      (reference.kind === 'form' ? 15 : 0) -
      (reference.width && reference.height && reference.width / reference.height < 0.65 ? 40 : 0)
    );
  };
  return references
    .filter(safeReference)
    .sort(
      (a, b) =>
        Number(isFullBodyReference(a)) - Number(isFullBodyReference(b)) || score(b) - score(a),
    );
}

File: internal/research/portraits.go
package research

import (
	"net/url"
	"regexp"
	"sort"

	"github.com/mardwerk/unit-generator/internal/unit"
)

var (
	fullBody     = regexp.MustCompile(`(?i)\b(full\s*body|whole\s*body|entire\s*body|turnaround|character sheet)\b`)
	portraitWord = regexp.MustCompile(`(?i)\b(portrait|headshot|close\s*up|bust)\b`)
	profileWord  = regexp.MustCompile(`(?i)\b(infobox|profile)\b`)
	animeWord    = regexp.MustCompile(`(?i)\banime\b`)
	separators   = regexp.MustCompile(`[_-]`)
)

func metadata(reference unit.VisualReference) string {
	text := reference.Caption + " " + reference.ID + " " + reference.URL
	if decoded, err := url.PathUnescape(text); err == nil {
		text = decoded
	}
	return separators.ReplaceAllString(text, " ")
}

// IsFullBody reports whether source labels identify a full-body reference.
// Labels describe coverage; no pixels are analyzed.
func IsFullBody(reference unit.VisualReference) bool {
	return fullBody.MatchString(metadata(reference))
}

// SafeReference reports whether both URLs are HTTP(S) without credentials.
func SafeReference(reference unit.VisualReference) bool {
	for _, value := range []string{reference.URL, reference.SourceURL} {
		parsed, err := url.Parse(value)
		if err != nil || (parsed.Scheme != "http" && parsed.Scheme != "https") || parsed.User != nil || parsed.Host == "" {
			return false
		}
	}
	return true
}

// RankedPortraits orders safe references for a portrait: portrait-labelled
// art first, full-body and action references last.
func RankedPortraits(references []unit.VisualReference) []unit.VisualReference {
	score := func(reference unit.VisualReference) int {
		text := metadata(reference)
		total := 0
		if reference.Kind == "appearance" {
			total += 15
		}
		if portraitWord.MatchString(text) {
			total += 80
		}
		if profileWord.MatchString(text) {
			total += 35
		}
		if animeWord.MatchString(text) {
			total += 10
		}
		if reference.Kind == "pose" {
			total -= 25
		}
		if reference.Kind == "form" {
			total -= 15
		}
		if reference.Width != nil && reference.Height != nil && *reference.Height > 0 && float64(*reference.Width)/float64(*reference.Height) < 0.65 {
			total -= 40
		}
		return total
	}
	var ranked []unit.VisualReference
	for _, reference := range references {
		if SafeReference(reference) {
			ranked = append(ranked, reference)
		}
	}
	sort.SliceStable(ranked, func(i, j int) bool {
		a, b := IsFullBody(ranked[i]), IsFullBody(ranked[j])
		if a != b {
			return !a
		}
		return score(ranked[i]) > score(ranked[j])
	})
	return ranked
}
